### PR TITLE
Handle recursive schema in OpenAPI with fallback

### PR DESCRIPTION
**Pull request by mproch**

### CODE REVIEW

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala [hunk 1]
```diff
@@ -48,23 +48,29 @@ case class SwaggerObject(elementType: Map[PropertyName, SwaggerTyped], additiona
...
+case object SwaggerRecursiveSchemaFallback extends SwaggerTyped
+
 object SwaggerTyped {
 
   @tailrec
-  def apply(schema: Schema[_], swaggerRefSchemas: SwaggerRefSchemas): SwaggerTyped = schema match {
-    case objectSchema: ObjectSchema => SwaggerObject(objectSchema, swaggerRefSchemas)
-    case mapSchema: MapSchema => SwaggerObject(mapSchema, swaggerRefSchemas)
-    case arraySchema: ArraySchema => SwaggerArray(arraySchema, swaggerRefSchemas)
+  def apply(schema: Schema[_], swaggerRefSchemas: SwaggerRefSchemas, usedSchemas: Set[String] = Set.empty): SwaggerTyped = schema match {
```
Comment (arkadius) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala:57]: Can we have two methods - one public, without usedSchemas and second private? Do you see some case when we would like to pass it from the upper level?

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala [hunk 2]
```diff
@@ -83,7 +89,7 @@ object SwaggerTyped {
     }
   }
 
-  private def swaggerUnion(schemas: java.util.List[Schema[_]], swaggerRefSchemas: SwaggerRefSchemas) = SwaggerUnion(schemas.asScala.map(SwaggerTyped(_, swaggerRefSchemas)).toList)
+  private def swaggerUnion(schemas: java.util.List[Schema[_]], swaggerRefSchemas: SwaggerRefSchemas) = SwaggerUnion(schemas.asScala.map(SwaggerTyped(_, swaggerRefSchemas, Set.empty)).toList)
```
Comment (arkadius) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala:92]: Why empty?

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala [hunk 3]
```diff
@@ -48,23 +48,29 @@ case class SwaggerObject(elementType: Map[PropertyName, SwaggerTyped], additiona
...
+    case objectSchema: ObjectSchema => SwaggerObject(objectSchema, swaggerRefSchemas, usedSchemas)
+    case mapSchema: MapSchema => SwaggerObject(mapSchema, swaggerRefSchemas, usedSchemas)
+    case arraySchema: ArraySchema => SwaggerArray(arraySchema, swaggerRefSchemas, usedSchemas)
     case _ => Option(schema.get$ref()) match {
+      //handle recursive schemas better
+      case Some(ref) if usedSchemas.contains(ref) =>
+        SwaggerRecursiveSchemaFallback
       case Some(ref) =>
-        SwaggerTyped(swaggerRefSchemas(ref), swaggerRefSchemas)
+        SwaggerTyped(swaggerRefSchemas(ref), swaggerRefSchemas, usedSchemas = usedSchemas + ref)
```
Comment (arkadius) [inline utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala:66]: Does it mean that we will find recursion only on "second" level?

File: utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala (hunk 1 above)
Comment (mproch) [reply utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/swagger/SwaggerTyped.scala]: :+1: add second method

**Review by arkadius — APPROVED** [review]

**Pull request merged**